pyface/ui/qt4/data_view/data_view_item_model.py

```diff
@@ -0,0 +1,235 @@
+        elif left == () and right == ():
+            # this is a row header change
+            # XXX this is currently not supported and not needed
+            pass
+        else:
+            for i, (top_row, bottom_row) in enumerate(zip(top, bottom)):
+                if top_row != bottom_row:
+                    break
+            top = top[:i+1]
+            bottom = bottom[:i+1]
+
+            top_left = self._to_model_index(top, left)
+            bottom_right = self._to_model_index(bottom, right)
+            self.dataChanged.emit(top_left, bottom_right)
+
+    # Structure methods
+
+    def parent(self, index):
+        if not index.isValid():
+            return QModelIndex()
+        parent = index.internalPointer()
+        if parent == Root:
+            return QModelIndex()
+
+        grandparent, row = self.model.index_manager.get_parent_and_row(parent)
+        return self.createIndex(row, 0, grandparent)
+
+    def index(self, row, column, parent):
+        if parent.isValid():
+            parent_index = self.model.index_manager.create_index(
+                parent.internalPointer(),
+                parent.row(),
+            )
+        else:
+            parent_index = Root
+        index = self.createIndex(row, column, parent_index)
+        return index
+
+    def rowCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            if self.model.can_have_children(row_index):
+                return self.model.get_row_count(row_index)
+        except Exception:
+            logger.exception("Error in rowCount")
+
+        return 0
+
+    def columnCount(self, index):
+        row_index = self._to_row_index(index)
+        try:
+            return self.model.get_column_count(row_index) + 1
+        except Exception:
+            logger.exception("Error in columnCount")
+
+    # Data methods
+
+    def flags(self, index):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+
+        flags = Qt.ItemIsEnabled | Qt.ItemIsSelectable
+        if is_qt5 and not self.model.can_have_children(row):
+            flags |= Qt.ItemNeverHasChildren
+
+        if value_type and value_type.can_edit(self.model, row, column):
+            flags |= Qt.ItemIsEditable
+
+        return flags
+
+    def data(self, index, role=Qt.DisplayRole):
+        row = self._to_row_index(index)
+        column = self._to_column_index(index)
+        value_type = self.model.get_value_type(row, column)
+        if not value_type:
+            return None
+
+        if role == Qt.DisplayRole:
+            if value_type.has_text(self.model, row, column):
+                return value_type.get_text(self.model, row, column)
+        elif role == Qt.EditRole:
+            if value_type.can_edit(self.model, row, column):
+                return value_type.get_editable(self.model, row, column)
+
```

Comment:
I think it will be a fairly stable API once we are done; but if it changes as noted before, we can use `hasattr` to check for existence of new methods before calling for backwards compatibility.

Enums don't solve the problem (plus the code is unnatural in Python - although maybe not if `match` is implemented :) ) because not you all of a sudden have code that is broken by getting an unexpected enum value.

One potential risk reduction technique may be to have names of the form `has_...`, `get_...`, `set_...` consistently, which reduces the chances of collisions later.  I don't want to go all the way to traits/enable-style magic method names, though.

Reply:
> Enums don't solve the problem (plus the code is unnatural in Python - although maybe not if match is implemented :) ) because you all of a sudden have code that is broken by getting an unexpected enum value.

The Qt way to solve this is to say upfront that implementations should return an invalid value if the enum given is not recognized. That needs to be decided early on as part of the design, I guess.

Agreed we don't want magic naming again.

Reply:
I'm happy with this as is.  If we were later to decide that an enum approach was the right way to do things, then we can always add that to the API with a default implementation that calls out the already existing methods.

Thinking about this, there is another design consideration around subclassing: it's easier to subclass and override the behaviour of one or two methods than to subclass and override a single method that takes an enum, handle the cases you want to (probably by calling a method ;)) and then calling super() to handle the remaining cases.

It's also easier to _test_.  If I override one of many methods, I just need to validate that my one method works as expected (say, that the colors are correct) and can trust that unrelated methods still work as advertised; while if I override a single method with an enum input, I need to test _everything_ in case I messed something up.

Qt is likely doing it that way because C++ is a much more rigid language.